File: Set 7/Palindrome Linked List/Palindrome_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class PalindromeList:
    def is_palindrome(self, head):
        values = []
        current = head
        while current:
            values.append(current.val)
            current = current.next

        left, right = 0, len(values) - 1
        while left < right and values[left] == values[right]:
            left += 1
            right -= 1
        return left >= right

    def construct_linked_list(self, values):
        if not values:
            return None
        head = ListNode(values[0])
        current = head
        for value in values[1:]:
            current.next = ListNode(value)
            current = current.next
        return head


t = int(input().strip())
for _ in range(t):
    values = list(map(int, input().strip().split()))
    solution = PalindromeList()
    head = solution.construct_linked_list(values)
    result = solution.is_palindrome(head)
    print("Palindrome" if result else "Not Palindrome")

File: Set 7/Palindrome Linked List/Palindrome_list.cpp
#include <iostream>
#include <vector>
#include <sstream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class PalindromeList {
public:
    bool isPalindrome(ListNode* head) {
        vector<int> list;
        while (head) {
            list.push_back(head->val);
            head = head->next;
        }

        int left = 0;
        int right = list.size() - 1;
        while (left < right && list[left] == list[right]) {
            left++;
            right--;
        }
        return left >= right;
    }

    static ListNode* constructLinkedList(const vector<int>& values) {
        if (values.empty()) return nullptr;
        ListNode* head = new ListNode(values[0]);
        ListNode* current = head;
        for (size_t i = 1; i < values.size(); i++) {
            current->next = new ListNode(values[i]);
            current = current->next;
        }
        return head;
    }
};

int main() {
    int t;
    cin >> t;
    cin.ignore(); 

    while (t--) {
        string line;
        getline(cin, line);
        stringstream ss(line);

        vector<int> values;
        int value;
        while (ss >> value) {
            values.push_back(value);
        }

        PalindromeList solution;
        ListNode* head = solution.constructLinkedList(values);
        bool result = solution.isPalindrome(head);

        cout << (result ? "Palindrome" : "Not Palindrome") << endl;

        while (head) {
            ListNode* temp = head;
            head = head->next;
            delete temp;
        }
    }

    return 0;
}
